src/lib.rs

```diff
@@ -0,0 +1,631 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use async_trait::async_trait;
+use prost::Message;
+use std::collections::HashMap;
+use std::sync::{
+    atomic::{AtomicU64, Ordering},
+    Arc, Mutex,
+};
+use std::time::Duration;
+use uprotocol_sdk::{
+    rpc::{RpcClient, RpcClientResult, RpcMapperError, RpcServer},
+    transport::{datamodel::UTransport, validator::Validators},
+    uprotocol::{
+        Data, UAttributes, UCode, UEntity, UMessage, UMessageType, UPayload, UPayloadFormat,
+        UPriority, UStatus, UUri, Uuid,
+    },
+    uri::{
+        serializer::{MicroUriSerializer, UriSerializer},
+        validator::UriValidator,
+    },
+};
+use zenoh::{
+    config::Config,
+    prelude::{r#async::*, Sample},
+    queryable::{Query, Queryable},
+    sample::AttachmentBuilder,
+    subscriber::Subscriber,
+};
+
+pub struct ZenohListener {}
+pub struct ULinkZenoh {
+    session: Arc<Session>,
+    subscriber_map: Arc<Mutex<HashMap<String, Subscriber<'static, ()>>>>,
+    queryable_map: Arc<Mutex<HashMap<String, Queryable<'static, ()>>>>,
+    query_map: Arc<Mutex<HashMap<String, Query>>>,
+    callback_counter: AtomicU64,
+}
+
+impl ULinkZenoh {
+    /// # Errors
+    /// Will return `Err` if unable to create Zenoh session
+    pub async fn new(config: Config) -> Result<ULinkZenoh, UStatus> {
+        let Ok(session) = zenoh::open(config).res().await else {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to open Zenoh session",
+            ));
+        };
+        Ok(ULinkZenoh {
+            session: Arc::new(session),
+            subscriber_map: Arc::new(Mutex::new(HashMap::new())),
+            queryable_map: Arc::new(Mutex::new(HashMap::new())),
+            query_map: Arc::new(Mutex::new(HashMap::new())),
+            callback_counter: AtomicU64::new(0),
+        })
+    }
+
+    fn to_zenoh_key_string(uri: &UUri) -> Result<String, UStatus> {
```

Comment:
IMHO it would be a good idea to prepend Zenoh keys specific to uProtocol with e.g. "up/" so that we namespace all the uProtocol communication happening over Zenoh to avoid collisions with any other systems on the same network using Zenoh.

Reply:
I don't think the key will conflict since we're using micro-form uuri.
However, I think it's a good way to let others know the key is from uProtocol instead of simply something unreadable to humans.